### ISSUE TITLE
Unify header validation across all codecs (H1, H2 and H3)

### ISSUE THREAD

**Issue by yanavlasov** [envoyproxy/envoy#10646]
Presently H1 and H2 (unsure about H3) codecs use header validation from nghttp2 library and have their own codec specific checks on top of it. The header validation code needs to be unified to make it easy to audit Envoy's header validation.

This change will include:
1. Adding header validation according to the HTTP spec.
1. Converting codec to use unified header validation facility.
1. Disabling header validation in nghtt2 library.

**Comment by mattklein123** [comment]
cc @danzh2010 for HTTP/3

**Comment by danzh2010** [comment]
QUICHE has its header validation done before handing it to codec. Is it suppose to be following the standard? If so QUICHE should already does so.

**Comment by mattklein123** [comment]
@danzh2010 the main thing we have to figure out here is how to share the various checks between all 3 codecs (QUICHE being one). Right now we have consistency issues just between 2, so when we add HTTP/3 the problem is even bigger. Specifically, we will have to implement header underscore ignore/drop in HTTP/3 to match HTTP/1 and HTTP/2 so this will be a forcing function to figure this out.

**Comment by danzh2010** [comment]
We can add extra validation steps after QUICHE delivers request header to codec, so it's something doable. But might not worth to check what the standard requires for a 2nd time.

**Comment by danzh2010** [comment]
>  Specifically, we will have to implement header underscore ignore/drop in HTTP/3 to match HTTP/1 and HTTP/2 so this will be a forcing function to figure this out.

As to underscore in header field, why wasn't it an http filter instead?

**Comment by mattklein123** [comment]
> As to underscore in header field, why wasn't it an http filter instead?

I think mostly for efficiency reasons as this is a potential attack vector, though I suppose in hindsight we could have done this at the HCM layer during decode headers. @yanavlasov @alyssawilk any thoughts here? We can change this later without any config implications though we would probably want to move the stats around.

**Comment by alyssawilk** [comment]
I think it was added as a commonly needed security fix - I think at some point we could refactor much of the HCM transformation into a shared filter, but in general most security work (making sure there's no smuggling, making sure there's no weird characters) go directly in the base classes.  Moving one out would be weird, and moving them all out would be dangerous :-P

**Comment by danzh2010** [comment]
What is the security concern for underscore in headers?

**Comment by mattklein123** [comment]
> What is the security concern for underscore in headers?

Some applications treat `_` and `-` as interchangeable (holdover from CGI days) and this opens up a fairly trivial security bypass vector in some cases. NGINX ignores headers with `_` in them by default for this reason.

> but in general most security work (making sure there's no smuggling, making sure there's no weird characters) go directly in the base classes.

+1, I don't think this stuff belongs in a filter, but I feel quite strongly that it belongs in common code, probably in the HCM in decodeHeaders. We need to make sure all of our HTTP level checks are done consistently across all codecs, and optimally not done twice (as nghttp2 does a bunch of them by default).

**Comment by antoniovicente** [comment]
This may be a topic for a separate feature request, but it would be good to run header validation relatively often in debug modes so we can detect cases where filters make modifications that break header invariants.  Re-running header checks after each filter that can modify headers and/or before serialization may help us detect issues early (e.i. in integration tests)

**Comment by alyssawilk** [comment]
I think this now goes beyond codec ingress into "untrusted data ingress"

We should probably be doing the trusted header checks after ext_proc and wasm to make sure
1) no x-envoy headers added by untrusted users
2) no spec violating or standard envoy guards (e.g. no Transfer-Encoding and Content-Length headers coexisting
3) no invalid input (bad characters, whitespace etc)

cc @PiotrSikora @mathetake @gbrail @htuch

**Comment by htuch** [comment]
Just ext_proc/wasm or all HTTP filters?

**Comment by antoniovicente** [comment]
For efficiency reasons it may make sense to have options to skip some of these checks after specific filters that are trusted enough to not introduce badness.  But effectively yes, it would be good to perform this sort of consistency check after each HTTP filter that can perform request or response header transformations.

**Comment by PiotrSikora** [comment]
Note that we would want to perform those checks inside the Wasm HTTP filter at the time when a change to headers happens, so that we could return an error back to the caller, not afterwards.

**Comment by gbrail** [comment]
It would be great to clarify this. The situation right now with ext_proc is
that I think that certain header manipulations will result in an assertion
failure that crashes Envoy, which really worries me! It'd be great if we
could instead have a standard library that returns a status as to whether a
particular header manipulation is valid, and configuration in ext_proc and
ext_authz that says what to do with an invalid manipulation.

For example, when Envoy gets a response back from an ext_proc processing
server making an invalid request to change a header, we have the choice of
ignoring the request to change the header (and logging or incrementing a
counter), or ignoring the ext_proc processing server for the rest of the
HTTP transaction (we do that if the server sends back response messages in
the wrong order). We also have the choice of letting the HTTP transaction
keep processing, or failing it with a 500 error, or bombing out Envoy I
suppose. All that should be configurable with a default that does not cause
Envoy to crash.

Finally, I think that we also need controls on what headers may be changed,
so that the proxy has some control over how much it wants to trust the
ext_proc processing server. https://github.com/envoyproxy/envoy/issues/14789
addresses these controls, and there's a PR in flight to try and implement
them, since the default that we have now is too restrictive for some use
cases.

**Comment by danzh2010** [comment]
cc @birenroy for oghttp2

**Comment by htuch** [comment]
@yanavlasov this is a nice summary - it might be easier to read as a table in which rows were equivalent-ish things, e.g. header keys, header values, host/:authority, and then protocol specific.

**Comment by yanavlasov** [comment]
@gbrail I think this is outside of the scope of this change. Envoy checks that required request or response headers are present and responds with an error status if the check fails. It should not ASSERT. If this is not the case I suggest opening a separate issue.
I think a lot of what you have listed is ext_proc specific. For clarification we could:

1. Document the required headers in the request or response header maps.
2. Add API for checking if a header is required for request or response to be proxied and the filter can you this API when deleting headers from the map.

If this sounds good to you, please open a separate issue to track this work.

**Comment by yanavlasov** [comment]
The unified header validation (UHF) component will do the following:
1. Validate header map for standard compliance. The validator will be protocol aware and do both validations common to all protocols and protocol specific validations.
1. Perform header map transformations before request is processed through the filter chain. These include:
   - Path normalization
   - TBD

The UHF component will be implemented as an extension, such that the default behavior can be overridden. The business case for it is transitioning operators of stacks with non-compliant behavior to Envoy based infrastructure.

Development Plan:

1. Capture header validation behavior in unit tests.
1. Define API (proto config and internal C++) for the UHF component.
1. Implement per-request flags that disable header validation for http-parser and nghttp2.
   - nghttp2 will be patched (the patch should be small and exist only until transition to QUICHE H/2 codec)
   - http-parser will be copied into Envoy’s repo as it is no longer maintained in its SoT repo.
1. Implement flag protected **name and value** validation **common** to all protocols and allow it some burn time. Protocol specific validation will still stay in codecs. Remove the runtime flag and start using unified header validation common to all protocols.
1. Implement flag protected path normalization in the UHF, burn time, flag removal.
1. Implement flag protected protocol specific header validation in the UHF, burn time, flag removal. These include:
   - H/1, H/0.9 request and response line validation
   - H/2 lowercase header names
   - H/2 pseudo headers, disallowed headers, TE: trailers
   - H/1 TE + CL checks
   - All other H/1 header checks codec_impl.cc

**NOTE:** steps 4 and 5 can be done in parallel.

**Comment by danzh2010** [comment]
> 3. Implement per-request flags that disable header validation for http-parser and nghttp2.

We are adding to QUICHE its own header validation logics. Would these verification preferred to be no-op in Envoy?

Do we have an estimated timeline for UHF? QUIC team is pushing hard to add its own header validation because of the existing gap between Envoy Http2 and Http3 validation. But if we are adding UHF soon, QUIC team can adjust the priority.

**Comment by yanavlasov** [comment]
@danzh2010  The plan is to start working on it early in Q1 22, as it also has impact on H/1 codec transition.

**Comment by ameily** [comment]
I'll be creating multiple issues to track each task for the larger goal of a unified header validation (uhv) component, based on the tech spec.

**Comment by ameily** [comment]
I've created the issues. Here is the rough order that I think we will work on them:

1. https://github.com/envoyproxy/envoy/issues/19749
2. https://github.com/envoyproxy/envoy/issues/19750
3. https://github.com/envoyproxy/envoy/issues/19751
4. https://github.com/envoyproxy/envoy/issues/19752
5. https://github.com/envoyproxy/envoy/issues/19753
6. https://github.com/envoyproxy/envoy/issues/19754
7. https://github.com/envoyproxy/envoy/issues/19755
8. https://github.com/envoyproxy/envoy/issues/19757

**Comment by yanavlasov** [comment]
The work on this issue is still in progress and is pending on completion of #29388

**Comment by huanghuangzym** [comment]
if http header  only have key, but no value . can 400 response?
like:
client (x-forward-protoc: "") ->  envoy (response 404)
add validate ?

**Comment by wdauchy** [comment]
@yanavlasov from 1.28 changelog I see
```
UHV: Introduced runtime flag envoy.reloadable_features.enable_universal_header_validator for toggling Universal Header Validator (UHV) on and off. The default value is off. This option is currently functional only when the ENVOY_ENABLE_UHV build flag is enabled. See https://github.com/envoyproxy/envoy/issues/10646 for more information about UHV.
```
I think it is still not the default to build with ENVOY_ENABLE_UHV flag.
I was curious what was the status of this feature?
